fix: GraphRAG does not respect .env #191

kun432 opened the pull request:
## Description

- Fixes #191
  - GraphRAG init command does not respect env vars `GRAPHRAG_LLM_MODEL` and `GRAPHRAG_EMBEDDING_MODEL` and always use gpt-4-turbo and text-embedding-3-small as default, which is current GraphRAG's behaviour.
  - If env vars above are set, This PR makes a copy of original GraphRAG's settings.yaml as backup and change settings.yaml to use the models defined by env vars in indexing phase.

NOTES: This is a **PARTIAL** fix. This fix works for OpenAI models only, not for Azure OpenAI and Ollama yet.

## Type of change

- [ ] New features (non-breaking change).
- [x] Bug fix (non-breaking change).
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected).

## Checklist

- [x] I have performed a self-review of my code.
- [ ] I have added thorough tests if it is a core feature.
- [x] There is a reference to the original bug report and related work.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [ ] The feature is well documented.

kun432 committed:
partially fix settings.yaml using env variabls for GraphRAG #191

kun432 commented:
test resutls in my end (installed from source):

.env

```ini
GRAPHRAG_API_KEY=XXXXXXXXXXXXXXXXXXXXX
GRAPHRAG_LLM_MODEL=gpt-4o-mini
GRAPHRAG_EMBEDDING_MODEL=text-embedding-3-small
```

creating index

![sc1](https://github.com/user-attachments/assets/0a45680d-716c-46f7-819e-9fbde63f9a33)

after creating index

```shell
$ find ktem_app_data -type f -name "settings.yaml*" -ls
 66474046      8 -rw-rw-r--   1 kun432   kun432       5198  9月  5 12:51 ktem_app_data/user_data/files/graphrag/81ea1c56-f74a-4fed-996e-164dd4e76759/settings.yaml.back
 66474039      4 -rw-rw-r--   1 kun432   kun432       1368  9月  5 12:51 ktem_app_data/user_data/files/graphrag/81ea1c56-f74a-4fed-996e-164dd4e76759/settings.yaml
```

settings.yaml

```yaml
encoding_model: cl100k_base
skip_workflows: []
llm:
  api_key: ${GRAPHRAG_API_KEY}
  type: openai_chat
  model: gpt-4o-mini
  model_supports_json: true
parallelization:
  stagger: 0.3
async_mode: threaded
embeddings:
  async_mode: threaded
  llm:
    api_key: ${GRAPHRAG_API_KEY}
    type: openai_embedding
    model: text-embedding-3-small
```

settings.yaml.back

```yaml

encoding_model: cl100k_base
skip_workflows: []
llm:
  api_key: ${GRAPHRAG_API_KEY}
  type: openai_chat # or azure_openai_chat
  model: gpt-4-turbo-preview
  model_supports_json: true # recommended if this is available for your model.
(snip)

embeddings:
  ## parallelization: override the global parallelization settings for embeddings
  async_mode: threaded # or asyncio
  # target: required # or all
  llm:
    api_key: \${GRAPHRAG_API_KEY}
    type: openai_embedding # or azure_openai_embedding
    model: text-embedding-3-small
(snip)
```

Also I checked:

- chat works using GraphRAG
- only the models defined in .env are used on OpenAI Usage dashboard.

oubeichen commented:
Have you tried testing if the `GRAPHRAG_EMBEDDING_MODEL` configuration is effective? Your are currently testing with the default model, which is `text-embedding-3-small`. I switched to a self-hosted LLM server and changed the `GRAPHRAG_EMBEDDING_MODEL` to something else, but the logs still show that `text-embedding-3-small` is being used.

kun432 commented:
in my end, GRAPHRAG_EMBEDDING_MODEL seems working.

.env

```ini
GRAPHRAG_API_KEY=XXXXXXXXXXXXXXXXXXXXX
GRAPHRAG_LLM_MODEL=gpt-4o-mini
GRAPHRAG_EMBEDDING_MODEL=text-embedding-3-large
```

![SS__2024-09-05_15_29_41](https://github.com/user-attachments/assets/cd81d37e-8575-496f-8e3e-7ce6e8a5e850)

generated settings.yaml

```
(snip)
embeddings:
  async_mode: threaded
  llm:
    api_key: ${GRAPHRAG_API_KEY}
    type: openai_embedding
    model: text-embedding-3-large
(snip)
```

what does this mean?

>I switched to a self-hosted LLM server and changed the GRAPHRAG_EMBEDDING_MODEL to something else,

Niko-La commented:
is this a fix for local ollama too? i am unable to get graphrag embeddings. and file embeddings are also breaking in the file tab but not from the Chat tab.

kun432 commented:
>NOTES: This is a **PARTIAL** fix. This fix works for OpenAI models only, not for Azure OpenAI and Ollama yet.

oubeichen commented:
> what does this mean?
> 
> > I switched to a self-hosted LLM server and changed the GRAPHRAG_EMBEDDING_MODEL to something else,


Currently, my project cannot use the official OpenAI services and relies on my own LLM and embedding services.

I understand now. In your previous response https://github.com/Cinnamon/kotaemon/issues/191#issuecomment-2326016380, your code was:

```python
graphrag_settings["embeddings"]["model"] = graphrag_embedding_model
```

So, the `GRAPHRAG_EMBEDDING_MODEL` environment variable was not being used. However, in this Pull Request, your code is:

```python
graphrag_settings["embeddings"]["llm"]["model"] = graphrag_embedding_model
```

I have corrected this part of the code, and it is now functioning properly. Anyway, thank you for the fix.

kun432 commented:
@oubeichen ahh, in my first fix in that issue, I focused on LLM (because costs me a lot) and I found it was not enough for embeddings when making PR.😅

kun432 commented:
I found a bad commit history in my PR so closed this one and will recreate a new one. 
sorry to bother you.
